### PR TITLE
Register scoped middleware in Application::routes()

### PR DESCRIPTION
**Pull request by garas**
Recent change to add CSRF middleware to application middleware queue made many users confused due to CSRF token errors.

CSRF, encrypted cookies, authentication etc middlewares are better uses as scoped middlewares.

I suggest to use `Application::routes()` to register (and group) scoped middlewares. This way all middlewares are still configured in single file, some in `routes()`, some in `middleware()`, but separated.

Add `applyMiddleware()` and short example about separate scopes in `routes.php`.

This PR separates app-wide and scoped middlewares, it might be easier to communicate about differences and enabling/disabling scoped middlewares. In my opinion, `routes.php` has more visibility than `Application.php` between beginner CakePHP developers.

### CODE REVIEW

File: src/Application.php [hunk 1]
```diff
@@ -57,6 +57,24 @@ public function bootstrap()
...
+     *
+     * @param \Cake\Routing\RouteBuilder $routes A route builder to add routes into.
+     * @return void
+     */
+    public function routes($routes)
+    {
+        // Register scoped middlewares for use in routes.php
+        $routes->registerMiddleware('csrf', new CsrfProtectionMiddleware([
+            'httpOnly' => true
+        ]));
```
Comment (markstory) [inline src/Application.php]: You can register this middleware in routes.php as well if you want everything in one place.

Comment (garas) [reply src/Application.php]: Some middlewares requires `Application` as `$this`, some like `SecurityHeaderMiddleware` or `BodyParserMiddleware` have somewhat complex configuration or logic and I think `config/` should avoid complex logic, code/logic should be in `src/`.

`routes.php` should be for configuring routes, not middlewares.

Comment (markstory) [reply src/Application.php]: Ok. The parent method is just a require for `config/routes.php` and the `$routes` variable is available there too.

File: config/routes.php [hunk 2]
```diff
@@ -47,6 +47,12 @@
 Router::defaultRouteClass(DashedRoute::class);
 
 Router::scope('/', function (RouteBuilder $routes) {
+    /**
+     * Apply a middlewares to the current route scope.
```
Comment (ravage84) [inline config/routes.php]: middlewares  -> middleware

File: config/routes.php [hunk 3]
```diff
@@ -47,6 +47,12 @@
...
+     * Requires middlewares to be registered via `Application::routes()` with `registerMiddleware()`
```
Comment (ravage84) [inline config/routes.php]: Plural of middleware is middleware. Confusing, but it is...

File: config/routes.php [hunk 4]
```diff
@@ -77,3 +83,14 @@
      */
     $routes->fallbacks(DashedRoute::class);
 });
+
+/**
+ * If you need different set of middlewares or none at all, open new scope and define routes there
```
Comment (ravage84) [inline config/routes.php]: middlewares -> middleware

File: src/Application.php [hunk 5]
```diff
@@ -57,6 +57,24 @@ public function bootstrap()
...
+     * Define the routes for an application.
+     *
+     * Use the provided RouteBuilder to define an application's routing, register scoped middlewares.
+     *
+     * @param \Cake\Routing\RouteBuilder $routes A route builder to add routes into.
+     * @return void
+     */
+    public function routes($routes)
+    {
+        // Register scoped middlewares for use in routes.php
```
Comment (ravage84) [inline src/Application.php]: middlewares  -> middleware

**Pull request merged**